### PR TITLE
Fix for FM-66

### PR DESCRIPTION
**Pull request by Tharakamora**
Implement FHIR condition resource mapper to get the FHIR Condition for OpenMrs Obs

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java [hunk 1]
```diff
@@ -13,15 +13,39 @@
...
+import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
 
 public class ObsConditionStrategy implements GenericConditionStrategy {
 
 	@Override
 	public Condition getConditionById(String uuid) {
-		return null;
+		String conceptsAsConditions = Context.getAdministrationService().getGlobalProperty("fhir.concepts.conditions");
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java]: move the global property name to a constants class

File: api/src/main/java/org/openmrs/module/fhir/api/impl/ConditionServiceImpl.java [hunk 2]
```diff
@@ -56,6 +57,10 @@ public Condition getCondition(String id) {
 	}
 
 	@Override
+	public Condition getConditionByObsId(String id) {
+		return ConditionStrategyUtil.getConditionStrategy().getConditionById(id);
+	}
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/ConditionServiceImpl.java]: need space after

File: api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java [hunk 3]
```diff
@@ -13,15 +13,39 @@
...
+		List<String> concepts = new ArrayList<String>();
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java]: cant we have integer array or something without converting this to string always

File: api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java [hunk 4]
```diff
@@ -33,4 +57,70 @@ public Condition getConditionById(String uuid) {
...
+		id.setValue(openMrsObs.getUuid());
+		fhirCondition.setId(id);
+
+		//Set patient reference
+		if (openMrsObs.getPerson().isPatient()) {
+			ResourceReferenceDt patient = FHIRUtils.buildPatientOrPersonResourceReference(openMrsObs.getPerson());
+			fhirCondition.setPatient(patient);
+		}
+		//Set Encounter
+
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java]: format comments

File: api/src/main/java/org/openmrs/module/fhir/api/impl/ConditionServiceImpl.java (hunk 2 above)
Comment (Tharakamora) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/ConditionServiceImpl.java]: Added a space after.

File: api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java (hunk 4 above)
Comment (Tharakamora) [inline api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java]: Still have to identify how the encounter and asserter is going to be mapped correctly. That's why these comments are there. Once they are done I will remove these comments.

**Comment by harsha89** [comment]
@Tharakamora can you write a test for this?

**Comment by Tharakamora** [comment]
@harsha89  I will surely write a test for this.

File: api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        String openmrsAdminPersonUuid = "dagh524f-27ce-4bb2-86d6-6d1d05312bd5";
+        String openmrsQuestionConceptUuid = "4a5048b1-cf85-4c64-9339-7cab41e5e364";
+        Date openmrsDateApplies = new Date();
+        Person admin = Context.getPersonService().getPersonByUuid(openmrsAdminPersonUuid);
+        Concept concept = Context.getConceptService().getConceptByUuid(openmrsQuestionConceptUuid);
+        Obs problemAddedObs = new Obs(admin, concept, openmrsDateApplies, null);
+        problemAddedObs.setValueNumeric(8d);
+        problemAddedObs = Context.getObsService().saveObs(problemAddedObs, null);
+        Condition fhirConditionForProblemAddedObs = getService().getConditionByObsId(problemAddedObs.getUuid());
+        assertNotNull(fhirConditionForProblemAddedObs);
```
Comment (harsha89) [inline api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java]: assert not null wouldnt not be enough can you add more assertions?

File: api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    public void runBeforeEachTest() throws Exception {
+        executeDataSet(OBS_INITIAL_DATA_XML);
+        executeDataSet(CONCEPT_CUSTOM_INITIAL_DATA_XML);
+        executeDataSet(PERSOM_INITIAL_DATA_XML);
+    }
+
+    @Test
+    public void getConditionByObsId_shouldReturnMatchingFHIRCondition() {
+        Context.getAdministrationService().saveGlobalProperty(new GlobalProperty(FHIRConstants
+                .CONCEPTS_CONVERTABLE_TO_CONDITIONS_STORED_AS_OBS, GLOBAL_PROPS_CONDITION_MAPPING_CONCEPT_ID_));
```
Comment (harsha89) [inline api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java]: GLOBAL_PROPS_CONDITION_MAPPING_CONCEPT_ID_ why there is last underscore?

File: api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        String openmrsAdminPersonUuid = "dagh524f-27ce-4bb2-86d6-6d1d05312bd5";
```
Comment (harsha89) [inline api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java]: use shortened meaningful names

File: api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java (hunk 4 above)
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/condition/ObsConditionStrategy.java]: Can we remove this now or still more work to be done?

File: api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java (hunk 5 above)
Comment (Tharakamora) [inline api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java]: added more assertions.

File: api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java (hunk 6 above)
Comment (Tharakamora) [inline api/src/test/java/org/openmrs/module/fhir/api/ConditionServiceTest.java]: That was a typo. Sorry about that. Fixed.

**Comment by harsha89** [comment]
looks good!

**Pull request merged**